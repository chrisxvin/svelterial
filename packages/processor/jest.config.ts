module.exports = {
  transform: {
    '^.+\\.tsx?$': 'ts-jest',
  },
  testMatch: ['**/tests/**/*.ts'],
  testPathIgnorePatterns: ['/tests/samples/'],
};
