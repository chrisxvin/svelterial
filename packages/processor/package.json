{
  "name": "@svelterial/processor",
  "version": "0.0.0",
  "description": "A svelte preprocessor for svelterial plugins.",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/TheComputerM/svelterial.git",
    "directory": "packages/processor"
  },
  "author": "TheComputerM",
  "homepage": "https://github.com/TheComputerM/svelterial#readme",
  "bugs": {
    "url": "https://github.com/TheComputerM/svelterial/issues"
  },
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "exports": {
    "require": "./dist/index.cjs",
    "import": "./dist/index.js",
    "node": "./dist/index.cjs"
  },
  "scripts": {
    "build": "pnpm types && rollup -c",
    "test": "jest",
    "types": "dts-bundle-generator -o dist/index.d.ts src/index.ts"
  },
  "files": [
    "dist/**"
  ],
  "keywords": [
    "svelte",
    "sveltejs",
    "preprocessor",
    "svelterial"
  ],
  "dependencies": {
    "deepmerge": "^4.2.2",
    "postcss": "^8.2.8",
    "sass": "^1.32.8"
  },
  "devDependencies": {
    "@rollup/plugin-commonjs": "^17.1.0",
    "@rollup/plugin-node-resolve": "^11.2.0",
    "@rollup/plugin-typescript": "^8.2.0",
    "@types/jest": "^26.0.21",
    "@types/node": "^14.14.35",
    "dts-bundle-generator": "^5.8.0",
    "is-plain-object": "^5.0.0",
    "jest": "^26.6.3",
    "rollup": "^2.42.3",
    "rollup-plugin-terser": "^7.0.2",
    "svelte": "^3.35.0",
    "ts-jest": "^26.5.4",
    "ts-node": "^9.1.1",
    "tslib": "^2.1.0",
    "typescript": "^4.2.3"
  },
  "types": "dist/index.d.ts"
}
