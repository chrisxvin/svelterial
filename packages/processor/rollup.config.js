import typescript from '@rollup/plugin-typescript';
import { nodeResolve } from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import { terser } from 'rollup-plugin-terser';
import pkg from './package.json';

/** @type {import('rollup').RollupOptions} */
export default {
  input: 'src/index.ts',
  output: [
    {
      file: pkg.main,
      format: 'cjs',
      exports: 'auto',
    },
    {
      file: pkg.module,
      format: 'esm',
      exports: 'auto',
    },
  ],
  plugins: [typescript(), nodeResolve(), commonjs(), terser()],
};
