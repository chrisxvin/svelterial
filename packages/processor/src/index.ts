import { isPlainObject } from 'is-plain-object';
import globalStyles from './global';
import variableTransformer from './variableTransformer';

const sass = require('sass');
const deepmerge = require('deepmerge');
const postcss = require('postcss');
const path = require('path');
const { readFileSync } = require('fs');

interface SvelterialPlugin {
  name: string;
  // eslint-disable-next-line no-unused-vars
  defaults: object | ((vars: object) => object);
}

interface ConfigOptions {
  cacheVariables: boolean;
}

export interface Config {
  variables?: object;
  options?: ConfigOptions;

  plugins?: SvelterialPlugin[];
}

interface Processor {
  content: string;
  attributes: {
    svelterial: boolean;
    name: string;
    src?: string;
  };
  filename: string;
}

const defaultConfig: Config = {
  variables: {},
  options: {
    cacheVariables: true,
  },
  plugins: [],
};

function parseVariables(config: Config) {
  const { variables, plugins } = config;
  return plugins.reduce((acc: object, plugin) => {
    const userConfig = variables?.[plugin.name] || {};
    if (isPlainObject(plugin.defaults)) {
      acc[plugin.name] = deepmerge(plugin.defaults, userConfig);
    } else {
      // TODO: fix this
      // @ts-ignore
      const pluginConfig = plugin.defaults(acc);
      acc[plugin.name] = deepmerge(pluginConfig, userConfig);
    }

    return acc;
  }, {});
}

export default (_config: Config) => {
  // For caching variables
  const cache = {
    config: null,
    variables: null,
    set: false,
  };

  return {
    async style({ content, attributes: info, filename }: Processor) {
      if (!info.svelterial) return null;

      let input: string = content;
      const dependencies: string[] = [];

      const config: Config = cache.config || deepmerge(defaultConfig, _config);

      if (config.options.cacheVariables) {
        if (!cache.set) {
          cache.config = deepmerge(defaultConfig, _config);
          cache.variables = parseVariables(config);
          cache.set = true;
        }
      }

      const variables = cache.variables || parseVariables(config);

      /**
       * Processing file if there is a 'src' attribute.
       */
      if (info.src) {
        const importedFilePath = path.resolve(path.dirname(filename), info.src);
        input = readFileSync(importedFilePath).toString();
        dependencies.push(importedFilePath);
      }

      const sassOutput = sass.renderSync({
        data: input,
        outFile: `${filename}.css`,
        outputStyle: 'compressed',
        sourceMap: true,
        omitSourceMapUrl: true,
        includePaths: [process.cwd()],
        importer: [
          (url: string) => {
            if (!url.startsWith('svelterial')) return null;

            // Name of the imported plugin.
            const name = url.split('/')[1];

            /**
             * Returns the SCSS variables that are to be used in the plugin.
             */
            return {
              contents: variableTransformer(variables[name] || {}),
            };
          },
        ],
      });

      dependencies.concat(sassOutput.stats.includedFiles);

      const output = await postcss([globalStyles()]).process(sassOutput.css.toString(), {
        from: filename,
        map: { prev: sassOutput.map.toString(), inline: false },
      });

      dependencies.concat(
        output.messages.reduce((acc, msg) => {
          if (msg.type !== 'dependency') return acc;
          acc.push(msg.file);

          return acc;
        }, []),
      );

      return {
        code: output.css,
        map: output.map,
        dependencies,
      };
    },
  };
};
