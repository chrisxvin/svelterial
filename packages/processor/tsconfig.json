{
  "compilerOptions": {
    "target": "ES2019",
    "module": "ES2020",
    "moduleResolution": "node",
    "noEmitOnError": true,
    "lib": ["ES2020"],
    "esModuleInterop": false,
    "allowSyntheticDefaultImports": true,
    "declaration": true,
    "emitDeclarationOnly": true,
    "outDir": "dist"
  },
  "include": ["src/index.ts"],
  "exclude": ["node_modules/**/*", "tests/**/*"]
}
