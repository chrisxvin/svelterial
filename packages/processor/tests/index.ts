import processor from '../src/index';

const { preprocess } = require('svelte/compiler');
const { readFileSync } = require('fs');
const path = require('path');

async function render(file: string) {
  const folder = path.join(__dirname, `./samples/${file}`);
  const config = await import(path.join(folder, '_config.ts'));
  process.chdir(folder);
  const source = readFileSync(path.join(folder, 'input.svelte')).toString();
  const { code } = await preprocess(source, processor(config), { filename: file });
  return code;
}

test.each([
  // Basic variable insertion.
  'basic',

  // SCSS like nesting.
  'nesting',

  // Checks if it can import file using '@import'.
  'import',

  // Checks if it can import file using 'src' attribute.
  'src',

  // Features such as loops and array variables.
  'arrays',

  // Features such as mixins.
  'mixins',

  // Features such as scss maps and object variables.
  'maps',

  // Test the variables option in the config file.
  'variables',

  // Test the settings option in the config file.
  'settings',
])('transform %s', async (name) => {
  expect(await render(name)).toMatchSnapshot();
});
