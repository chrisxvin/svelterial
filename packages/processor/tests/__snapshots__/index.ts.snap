// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform @import 1`] = `
"<style svelterial>:global(h1){color:red}</style>
"
`;

exports[`transform @mixin 1`] = `
"<style svelterial>:global(h1),:global(h2),:global(h3){color:red;font-size:4em}</style>
"
`;

exports[`transform @use 1`] = `
"<style svelterial>:global(h1){color:red}</style>
"
`;

exports[`transform arrays 1`] = `
"<style svelterial>:global(.width-1){width:10em}:global(.width-2){width:20em}:global(.width-3){width:30em}:global(.width-4){width:40em}:global(.width-5){width:50em}:global(.height-1){height:10em}:global(.height-3){height:30em}:global(.height-5){height:50em}:global(.color-red){color:red}:global(.color-blue){color:blue}:global(.color-green){color:green}</style>
"
`;

exports[`transform basic 1`] = `
"<style svelterial>:global(h1){color:red}</style>
"
`;

exports[`transform import 1`] = `
"<style svelterial>:global(h1){color:red}</style>
"
`;

exports[`transform maps 1`] = `
"<style svelterial>:global(h1){font-size:200px}</style>
"
`;

exports[`transform mixins 1`] = `
"<style svelterial>:global(h1),:global(h2),:global(h3){color:red;font-size:4em}</style>
"
`;

exports[`transform multiple-plugins 1`] = `
"<style svelterial>:global(h1){font-size:16px;color:red;font-family:\\"Roboto\\"}</style>
"
`;

exports[`transform nesting 1`] = `
"<style svelterial>:global(div h1){font-size:16px}</style>
"
`;

exports[`transform settings 1`] = `
"<style svelterial>:global(h1){color:red}</style>
"
`;

exports[`transform src 1`] = `
"<style svelterial src=\\"./input.scss\\">:global(div h1){font-size:16}</style>
"
`;

exports[`transform variables 1`] = `
"<style svelterial>:global(h1){color:blue}</style>
"
`;
